kratos/utilities/intersection_utilities.h

```diff
@@ -233,6 +234,89 @@ class IntersectionUtilities
         return true;
     }
 
+    /**
+     * @brief Calculates the line to line intersection (shortest line). If line is length 0, it is considered a point and therefore there is intersection
+     * @details Calculate the line segment PaPb that is the shortest route between two lines P1P2 and P3P4. Calculate also the values of mua and mub where
+     *    Pa = P1 + mua (P2 - P1)
+     *    Pb = P3 + mub (P4 - P3)
+     *    http://paulbourke.net/geometry/pointlineplane/
+     * @param rSegment1 The first segment
+     * @param rSegment2 The second segment
+     * @tparam TGeometryType The geometry type
+     * @return Return empty points array if no solution exists. Otherwise returns the line intersection points array
+     */
+    template<class TGeometryType>
+    static PointerVector<Point> ComputeShortestLineBetweenTwoLines(
+        const TGeometryType& rSegment1,
+        const TGeometryType& rSegment2
+        )
+    {
+        // Zero tolerance
+        const double zero_tolerance = std::numeric_limits<double>::epsilon();
+
+        // Check geometry type
+        const auto geom_type = rSegment1.GetGeometryType();
+        KRATOS_ERROR_IF((geom_type != GeometryData::KratosGeometryType::Kratos_Line2D2) && geom_type != GeometryData::KratosGeometryType::Kratos_Line3D2) << "The geometry type is not correct, it is suppossed to be a linear line" << std::endl;
```

Comment:
And if we do an `if consexpr` to do the check "at compile time"?

Reply:
But you can't use the type from the template argument?

Reply:
No, the template argument only ensures that you can pass a geometry of nodes or points or any derived point class, if you enforce manually setting Line3D2<Point> or Line2D2<Point> for example, that will not work anyway because we don't know in compilation time that type, and `GetGeometryType` is not a static method. If you want to do that, that it will practical I admit, we need to add a static method to every geometry time returning it's geometry type. We will be using static polimorphism and I guess this will not work well with the dynamic polimorphism we use for geometries. @matekelemen do you think so?

Reply:
@loumalouomega  is unfortunately right, it can't be `constexpr` because `Geometry::GetGeometryType` is `virtual` (dynamic poly), which is a purely runtime abstraction. You do have some other options though.

You could use [`std::enable_if`](https://en.cppreference.com/w/cpp/types/enable_if) to restrict the types of geometries `ComputeShortestLineBetweenTwoLines` can be called with. Here's an [example on godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOykrgAyeAyYAHI%2BAEaYxCCSAJykAA6oCoRODB7evgGp6ZkCoeFRLLHxSbaY9o4CQgRMxAQ5Pn6BdpgOWQ1NBCWRMXEJyQqNza15HeP9YYPlw0kAlLaoXsTI7BzmAMxhyN5YANQmO25jxGHAp9gmGgCCu/uHmCdnTheYrDd3jw%2B/Fy8DiOAHFMKg2ARiABPX4mfxWB5HZFHABueGaXjERzG6BAIAuVyOwEwBAAKtCUpgIEsjmgGGMTgjiKT1gwTmYzGCIaSYeYzKcrP4ACInf4iwVw/4PQHAgYaI4gUHgyF88WI%2B4onEEPEEqFEknkynU2n0xmoVFxS7HeEWFkENkcszy/mSiXi4WS6V/e6yghHAZcRXKnlQ2Hq35a3H4wlGYmkilUml0gTmy3Ea2vW32x38wOunZCz0er0%2BgFQoH%2BoQ8gDyBAQcW5quhwabvPDjwRkZR0b1lzjhsTJpTDP9FqteBtzNZxHZ/OrbDrDeIbbDBaLcIlhalPoImBYKQMe7ebm7WvPdIMCgURzJq5hpDPF57OvxjCY0XoAH08FQvwRTnOV8QDwBQvwUVhMC/VFALvFV21IF0dmwJlT38NxtV1UDwMg6DYPvaFELmLgblIaJUE8G43lFCtMCfc8fgeVFUEnI50FQBdSQQK4awYe4BggM1/Tg0M%2BTMAA2I5iC/ElROhJZNw1KNgLQLx/UAwCpJk%2BCwwAOkHY1kw0s5MLfBh0FoUt4WLMsHjCf0WCYMIaUUp8CPjOTSy1eUjloOYvw0LyUUDXz/JI7ckRRTil0bHSYSOVB6ziKzIuRAB6NL2M4%2BseL4gTZObBTCyODKT1OT0dlFK0SCVBhUCORyCGQbi4yoLwGG6AQnyy0MWuAXj%2BLmCA/PCAKiqUlEeshPqBoEkaoK4canwyqauNywbwggRLl3Gkq0rKyryqOar4iOOqGqYJq%2BqONqOrqBhN2FDgVloTgAFZeD8DgtFIVBOFPSxrBxNYNizMwdh4UgCE0Z6VgbJgsHiGlSAAaxAN6uF0nZJHEgAODQuGxvGCZ2cT9E4SReBYCQNA0Ugvp%2Bv6OF4BQQDp6Hvue0g4FgGBEBATBVC6NSSHISgmmABRlEMGohAQVAAHcvshtADzoS6sml8JaDlxWGd4VWUjoYZgC4TlSEN43iBrNTdaVmGgiF5B7mISXOF4QWugafAvt4fhBBEMR2CkGRBEUFR1E50hdC4fRDGMaxrH0PBolZyAVlQFJ7tZjgAFpcXK0xAcsLh/BZkHNj0XEwi12X5ft7heChTAtkhhXiCYFJOB4F73s%2Bh2mewJ2ReII5VFx8Tc/EyRiWQZAjjN3SzCOCAAasSxSCOXBCBIDkIaWJuYaWOGvkRyhe44SnSGpnZ/F03Gdkf/x8dxyQzfEt6J/pgf3dsNmoaPtzPmEAkCe2QCPMWEAJZSxljreuysDYQiNkeTWsC7b6wtkgq2KADBGC4G9WmmC1b0GtrbeBDswEuzdszR2XsdRhF/v7YQohxAhyYeHNQDtdA7DjkYFAicN5%2BVTvADOWcsg53zjqQuAiLCl3LusSusdq4MFrnAvW3cm4slbrwduncNEXw%2Bt/KOg9h5EFHuPSe09LzxwXm9XSGh7ErzXknLe%2BAzF70WofTmx9SDwzPsjV6l8qYgAhvYh%2BEMNCJDev4RIpN/CxwwUzFm/8OZaBWDzfmYCIEUCga7GB2t0EaKIcgjWAhVGFKjpbEhIBUTIBSCkaCXBEhfiPJgMYX4LHFKtjbYQ5Co6ULyb/MB3sGE0KYYHVh0h2FKE4VHXQZheEJ2LjYIRadkaZ2zpwNK0Yi7r1kWXVYCjg62HoSotBfTIbN20aQXRXdG4GP7sYzgQ9hbuIsQ1BQtS0RNN0q0xkq8ZGb23u43YniAHeJPgjYYASKbBNCRocJXBInRNieJeJRjGa/1ZuzI%2BKw0aSFxrpD%2BENEgTzeiSieZNAk7EeZimhqTYbkw4GYWlvAkngrSSsdMGRnCSCAA%3D%3D). The advantage of this approach is that you'll get errors at compile time if you try to misuse it, but the drawback is that you either have to store the derived types directly (`Line2D2`s or `Line3D2`s), or `dynamic_cast` the base class pointers before you invoke this function on them. I'm not a fan of any type of casting and it's unlikely to be any faster than checking `GetGeometryType`, but at least the user has to go out of their way to misuse this function, so it's a bit safer.

The other solution, my preferred one, is to just loop through all edges in the geometry.

Reply:
OK. OFC you're both right. 

In any case I think that in this case the good approach would be to override the method with `Line2D2` and `Line3D2` being the node type the template argument. In that way you don't need to do the type check and you can optimize the 2D implementation by suppressing all the z-component operations.

Reply:
The only way I could do it was with explicit specialization, explicit instantiation was giving me compilation errors I could not solve